### ISSUE TITLE
Limit namespaces using RBAC

### ISSUE THREAD

**Issue by mlbiam** [kiali/kiali#5185]
When running Kiali in a multi-tenant environment, where there's a single kiali for the cluster, Kiali will attempt to load resources from namespaces that the user does not have access to.  On small clusters with a handful of namespaces, this isn't an issue.  On clusters with hundreds of namespaces the cluster gets bogged down to the point of not being usable.

When `deployment.accessible_namespaces` is `[**]`, kiali should loop through each namespace and attempt to retrieve the individual namespace object.  If the namespace is not accessible, it isn't returned.  This function is in https://github.com/kiali/kiali/blob/8191532af24f5f93c9534eada25d0557bbc0996a/business/namespaces.go#L62.  It's already managed this way of namespaces are explicitly listed in `deployment.accessible_namespaces`.  This feature update would do the same thing when `deployment.accessible_namespaces` is `[**]`.

I'm happy to submit a PR if the feature will be accepted

**Comment by lucasponce** [comment]
Thanks, @mlbiam for bringing this. It's a good moment as the project is focused on improving the experience on scalability scenarios.

I think that what you propose is interesting but it should be synced together with a broader revision of the flow.

This issue is also related https://github.com/kiali/kiali/issues/4678

In short, Kiali was designed to not use cluster permissions, with that limitation, yes, some way to define the namespaces is necessary (i.e. accessible_namespaces or similar).

Due to the lack of cluster permissions, when Kiali access to the k8s API using informers, it creates an informer per namespace, in small to mid namespaces scenarios, that's ok, but when users have a large number of namespaces that could create some extra usage of memory and network (also, creating an informer per user/token is not possible, so it uses the services account for that).

So, I think that for large namespaces scenarios, kiali could operate with cluster permissions to create single informers per the cluster, and additionally, it should control which namespaces a user can access (also, querying RBAC is a heavy query, so this operation should be done in a first step and cached somehow; this was done before in Kiali when looking in the overview page, there were RBAC queries per namespace and the penalty performance was pretty visible, so we refactor that logic).

In short, is a good topic but it would have more implications that would need some revision of the design to see how to connect these ideas to have a whole picture.

@mlbiam could interesting if you could share some anonymous telemetry of your cluster, like a number of namespaces, workloads, services, etc. That would help us a lot with these steps.

Let's keep discussing the design; when all ideas look ok then we can prototype some PRs with it.

**Comment by jmazzitelli** [comment]
I will say for multi-tenant scenarios, the suggested/recommended solution is to use [accessible_namespaces](https://kiali.io/docs/configuration/namespace-management/) (i.e. do NOT use accessible namespaces of `['**']`) - so one Kiali instance per tenant. That's how it is configured and working with OpenShift Service Mesh (which is multi-tenant by design). (You can install multiple Kialis per control plane - you can put them either in different namespaces or in the same namespace with different `spec.deployment.instance_name` values in their respective Kiali CRs).

Note that accessible_namespaces supports regex, so it is possible to limit the namespaces by using regex assuming the tenant namespaces follow patterns.. So e.g.

east-datacenter tenant:
```yaml
spec:
  deployment:
    accessible_namespaces:
    - east-datacenter-.*
```

west-dc tenant:
```yaml
spec:
  deployment:
    accessible_namespaces:
    - west-dc.*
```

So that is the "workaround" for this issue (and is, in fact, not a workaround but the recommended solution).

**Comment by lucasponce** [comment]
@nick the Kubernetes API is really badly designed in terms of users.
The namespaces API (if it's not changed in recent versions) requires cluster admins, so, you'd need to ask if you can access or not.

That's one of the reasons why OpenShift introduced the "Projects" API, which allows having that query (i.e. give me the projects I have access to).

So, unless there is a change in the main Kubernetes API, I don't see a way to do this in a clean way, but what Kiali has.

The `SelfSubjectAccessReview` is a very heavy API that can't be cached and we had that for the same scenario: "give me if I have access to this to render in the UI what I can do", but we changed that as well.

I'm interested in the discussion to learn if there was any significant change in the Kubernetes API in case we can adopt some new API but, yes, otherwise the `accessible_namespaces` should continue to be the way.

Perhaps, what @mlbiam would like to do is something that the operator's logic is doing: to collect the new namespaces that a user can see and update the accessible_namespaces accordingly.

But here we are touching on the logic that may depend on specific platforms and we'd need to see if any potential proposal can co-exist with current logic.

**Comment by mlbiam** [comment]
To @lucasponce points:

> The namespaces API (if it's not changed in recent versions) requires cluster admins, so, you'd need to ask if you can access or not.

No, to list namespaces you need `list` on the namespace API.  Not a cluster-admin.  

> That's one of the reasons why OpenShift introduced the "Projects" API, which allows having that query (i.e. give me the projects I have access to).

This is why I suggested making the list of users namespaces available via an aggregated API a possibility instead of querying the namespaces API directly.  Even though Kubernetes doesn't have its own built in way to know "what namespaces a user has access to", there's nothing that stops me from building that.  For instance, OpenUnison's Namespace as a Service knows what namespaces every user has access to without having to poll the API server.  If Kiali could poll OpenUnison via an aggregated API then it could leverage that knowledge to get the right list of namespaces securely.

Being able to assert which namespaces are available via the `userinfo` object in an extra field provides a similar function without having to poll even an aggregate API.  It makes it much easier for platform designers to provide a better UX without having to customize Kiali.

> I'm interested in the discussion to learn if there was any significant change in the Kubernetes API in case we can adopt some new API but, yes, otherwise the accessible_namespaces should continue to be the way.

forcing Kiali to a single instance per tenant doesn't scale well from a management perspective.  If we took this approach, each new tenant would mean:

"Day 1"

1. Deploy kiali object with the appropriate `accessible_namespaces`
2. Deploy a new proxy & SSO trust to manage authentication
3. Deploy new portal links so that users know they have access to it

"Day 2"
1. Update every instance of Kiali as new containers come out
2. update `accessible_namespaces` whenever a tenant's namespace change

The code to automate that is pretty intense, vs the code to provide a claim or an aggregate api?  Especially since Kiali has already centered the code needed for this into a single function.

> Perhaps, what @mlbiam would like to do is something that the operator's logic is doing: to collect the new namespaces that a user can see and update the accessible_namespaces accordingly.

No, I don't want to have to start messing with the current static logic because it doesn't scale from a management perspective, and any work done to generate the list of namespaces should only be done once in a session.  Even with hundreds of namespaces that woiuld need to be called in the "brute force approach" you're talking a few extra milliseconds.

To @nrfox points:

| Storing the list of namespaces the user has access to as an extra claim? It seems like there would be limits to how much info you can store there.

I don't know of any limit.  Maybe the 1mb etcd value limit?

| That's interesting although maybe less user friendly than option 2. Whatever aggregated API you are calling out to would probably need to do something similar to option 2 anyways so perhaps you are just shifting the burden onto something else. I'm not sure how OpenShift does this.

If you built a generic aggregate api?  yes, you'd need the same brute force method of just checking each one.  But if you're building a platform that manages this data, an aggregate API can be much simpler.

I suggested implementing all three methods because different platforms will want to supply that data in different ways.  This additional flexibility shouldn't have a major impact on Kiali but keeps users from having to resort to proprietary UIs that replicate kiali and makes it easier to work with the standard open source tools that are available.

**Comment by mlbiam** [comment]
@lucasponce some numbers for you:

42 clusters

est per cluster:
100 nodes / 16x128
300 namespaces
600 users
60 tenants

These are all on-prem clusters run as Anthos

**Comment by lucasponce** [comment]
Thanks, @mlbiam for the telemetry. That's really helpful.

> No, to list namespaces you need to list on the namespace API. Not a cluster-admin.

I think my response was also including the use of informers to access the Kubernetes API, I'm also including that in my thoughts about this, and for that list + watch are needed and if we want to avoid iterating on namespaces, it would need to watch on the cluster.

That's why I think this discussion is also related to https://github.com/kiali/kiali/issues/4678, as, in such a large cluster, the kiali backend should operate with the cluster rights to avoid creating 300x20 informers (every informer is an established connection watching for changes).

And then yes, it would require a way where Kiali can get what namespaces a user can access too.

**Comment by nrfox** [comment]
> I think my response was also including the use of informers to access the Kubernetes API, I'm also including that in my thoughts about this, and for that list + watch are needed and if we want to avoid iterating on namespaces, it would need to watch on the cluster.

@lucasponce Having less informers might help here or it might not. IMO this is less about cluster scoped vs. namespace scoped resources for what **Kiali** has access to and more about how Kiali gathers resources for what the **user** has access to. Even if Kiali used cluster-scoped informers, when `accessible_namespaces == **`, Kiali would either assume that each user has access to all namespaces (what it does today in non-openshift environments) or check the user's access for each namespace before gathering the resources (what this issue is proposing and how `accessible_namspaces != **` works today) from each namespace.

@mlbiam a few followup questions... where are you experiencing the slowdown? On a particular page in Kiali e.g. the "graph" page or the "overview" page? From a particular endpoint e.g. `/tls` or `/workloads`? How did you determine that Kiali loading resources for namespaces that users don't have access to was the cause of the slowdown? Thanks for providing info and doing some digging here.

**Comment by jshaughn** [comment]
This seems like an interesting issue but has not seen recent activity.  Since the last comment Kiali has implemented the cluster-scoped cache for the `accessible_namespaces == **` scenario, although as mentioned above that may not be what's needed here.

@mlbiam Any further comments to the question above?  @nrfox Anything you think we should immediately pursue here?

**Comment by jshaughn** [comment]
Closing for inactivity, but I believe this has been addressed in several ways, including the support discover selectors and major revisions to caching.
